File: llm_written/gpt-5/sccs/python/algo_parallel.py
from __future__ import annotations
from typing import List, Set, Tuple, Iterable, Optional
import os
from concurrent.futures import ProcessPoolExecutor

# ---------------------- Baseline (sequential, as provided) ----------------------
class GraphSequential:
    def __init__(self, v: int, verbose: bool = False) -> None:
        self.V: int = v
        self.adj: List[List[int]] = [[] for _ in range(v)]
        self.rev_adj: List[List[int]] = [[] for _ in range(v)]
        self.verbose = verbose

    def add_edge(self, v: int, w: int) -> None:
        self.adj[v].append(w)
        self.rev_adj[w].append(v)
    
    def tarjan_dfs(self,
        u: int,
        disc: List[int],
        low: List[int],
        stack: List[int],
        in_stack: List[bool],
        time_ref: List[int],
        scc_list: List[List[int]],
    ) -> None:
        time_ref[0] += 1
        disc[u] = low[u] = time_ref[0]
        stack.append(u)
        in_stack[u] = True                

        for v in self.adj[u]:
            if disc[v] == -1:
                self.tarjan_dfs(v, disc, low, stack, in_stack, time_ref, scc_list)
                low[u] = min(low[u], low[v])
            elif in_stack[v]:
                low[u] = min(low[u], disc[v])
        
        if low[u] == disc[u]:
            scc: List[int] = []
            while True:
                w = stack.pop()
                in_stack[w] = False
                scc.append(w)
                if w == u:
                    break
            scc_list.append(scc)

    
    def find_sccs(self) -> List[List[int]]:
        disc = [-1] * self.V
        low = [-1] * self.V
        in_stack = [False] * self.V
        stack: List[int] = []
        scc_list: List[List[int]] = []
        time_ref = [0]  # acts like 'ref int time'

        for i in range(self.V):
            if disc[i] == -1:
                self.tarjan_dfs(i, disc, low, stack, in_stack, time_ref, scc_list)

        return scc_list

    # Minimal SCC Edge Reduction (O(V + E))
    def minimize_edges_in_scc(self, scc: List[int]) -> List[Tuple[int, int]]:
        nodes: Set[int] = set(scc)
        essential_edges: List[Tuple[int, int]] = []

        # Step 1: forward spanning tree using DFS
        forward_tree = self.build_spanning_tree(scc[0], self.adj, nodes)

        # Step 2: reverse spanning tree using DFS on reversed graph
        reverse_tree = self.build_spanning_tree(scc[0], self.rev_adj, nodes)

        # Step 3: merge both trees (each edge appears at most twice)
        essential_edges.extend(forward_tree)
        essential_edges.extend(reverse_tree)

        return essential_edges

    def build_spanning_tree(
        self, start: int, graph: List[List[int]], nodes: Set[int]
    ) -> Set[Tuple[int, int]]:
        spanning_tree: Set[Tuple[int, int]] = set()
        visited: Set[int] = set()
        stack: List[int] = [start]
        visited.add(start)
        
        while stack:
            node = stack.pop()
            for neighbor in graph[node]:
                if neighbor in nodes and neighbor not in visited:
                    spanning_tree.add((node, neighbor))
                    visited.add(neighbor)
                    stack.append(neighbor)
        
        return spanning_tree

    def reduce_edges(self) -> List[Tuple[int, int]]:
        SCCs = self.find_sccs()
        if self.verbose:
            print(f"Found {len(SCCs)} SCC(s).")

        reduced_edges: List[Tuple[int, int]] = []
        for scc in SCCs:
            min_edges = self.minimize_edges_in_scc(scc)
            reduced_edges.extend(min_edges)

        if self.verbose:
            print(f"Reduced SCC edges: {len(reduced_edges)}")
        return reduced_edges


# ---------------------- Parallel, deterministic implementation ----------------------
# Globals used by worker processes (set via initializer)
_G_ADJ: Optional[List[List[int]]] = None
_G_REV: Optional[List[List[int]]] = None
_G_V: Optional[int] = None

def _init_worker_graph(adj: List[List[int]], rev: List[List[int]]) -> None:
    global _G_ADJ, _G_REV, _G_V
    _G_ADJ = adj
    _G_REV = rev
    _G_V = len(adj)


def _worker_minimize_scc(scc: List[int]) -> List[Tuple[int, int]]:
    # Runs inside a worker process; uses globals for adjacency, calls baseline to guarantee exact parity
    if _G_ADJ is None or _G_REV is None or _G_V is None:
        raise RuntimeError("Worker graph not initialized")
    g = GraphSequential(_G_V)
    g.adj = _G_ADJ
    g.rev_adj = _G_REV
    return g.minimize_edges_in_scc(scc)


def canonicalize_edges(edges: Iterable[Tuple[int, int]]) -> List[Tuple[int, int]]:
    # Sort and deduplicate deterministically
    seen = set()
    out: List[Tuple[int, int]] = []
    for e in edges:
        if e not in seen:
            seen.add(e)
            out.append(e)
    out.sort()
    return out


class GraphParallel:
    def __init__(self, v: int, verbose: bool = False, workers: Optional[int] = None, small_threshold: int = 2) -> None:
        self.V: int = v
        self.adj: List[List[int]] = [[] for _ in range(v)]
        self.rev_adj: List[List[int]] = [[] for _ in range(v)]
        self.verbose = verbose
        self.workers = workers if workers and workers > 0 else (os.cpu_count() or 1)
        # If number of SCCs <= small_threshold, run sequential per-SCC to avoid overhead
        self.small_threshold = max(0, small_threshold)

    def add_edge(self, v: int, w: int) -> None:
        self.adj[v].append(w)
        self.rev_adj[w].append(v)

    def find_sccs(self) -> List[List[int]]:
        return self._find_sccs_via_baseline()

    # We delegate SCC computation to the sequential baseline to ensure identical SCC ordering
    def _find_sccs_via_baseline(self) -> List[List[int]]:
        g = GraphSequential(self.V)
        g.adj = [lst[:] for lst in self.adj]
        g.rev_adj = [lst[:] for lst in self.rev_adj]
        return g.find_sccs()

    def _minimize_edges_in_scc_seq(self, scc: List[int]) -> List[Tuple[int, int]]:
        # Use baseline implementation to guarantee identical choice of edges
        g = GraphSequential(self.V)
        g.adj = self.adj
        g.rev_adj = self.rev_adj
        return g.minimize_edges_in_scc(scc)

    def reduce_edges(self) -> List[Tuple[int, int]]:
        # Compute SCCs using the baseline path to match scc[0] choice exactly
        SCCs = self._find_sccs_via_baseline()
        if self.verbose:
            print(f"Found {len(SCCs)} SCC(s). Using up to {self.workers} workers.")
        if len(SCCs) <= self.small_threshold or self.workers == 1:
            combined: List[Tuple[int, int]] = []
            for scc in SCCs:
                combined.extend(self._minimize_edges_in_scc_seq(scc))
            return canonicalize_edges(combined)
        # Parallel per-SCC phase with fixed order combine
        with ProcessPoolExecutor(max_workers=self.workers, initializer=_init_worker_graph, initargs=(self.adj, self.rev_adj)) as ex:
            # map preserves input order; chunksize=1 to avoid grouping across SCCs
            results = list(ex.map(_worker_minimize_scc, SCCs, chunksize=1))
        combined: List[Tuple[int, int]] = []
        for part in results:  # fixed combine order
            combined.extend(part)
        return canonicalize_edges(combined)

File: llm_written/gpt-5/sccs/rust/algo_parallel.rs
use std::collections::HashSet;

pub mod par {
    use super::*;
    use rayon::prelude::*;

    pub struct Graph {
        v: usize,
        adj: Vec<Vec<usize>>,
        rev_adj: Vec<Vec<usize>>,
        pub verbose: bool,
    }

    impl Graph {
        pub fn new(v: usize) -> Self { Self::with_verbose(v, false) }
        pub fn with_verbose(v: usize, verbose: bool) -> Self {
            Graph { v, adj: vec![Vec::new(); v], rev_adj: vec![Vec::new(); v], verbose }
        }
        pub fn add_edge(&mut self, v: usize, w: usize) {
            self.adj[v].push(w);
            self.rev_adj[w].push(v);
        }

        // Tarjan's SCC DFS (sequential)
        fn tarjan_dfs(
            &self,
            u: usize,
            disc: &mut [i32],
            low: &mut [i32],
            stack: &mut Vec<usize>,
            in_stack: &mut [bool],
            timer: &mut i32,
            scc_list: &mut Vec<Vec<usize>>,
        ) {
            *timer += 1;
            disc[u] = *timer;
            low[u] = *timer;
            stack.push(u);
            in_stack[u] = true;

            for &v in &self.adj[u] {
                if disc[v] == -1 {
                    self.tarjan_dfs(v, disc, low, stack, in_stack, timer, scc_list);
                    low[u] = low[u].min(low[v]);
                } else if in_stack[v] { low[u] = low[u].min(disc[v]); }
            }

            if low[u] == disc[u] {
                let mut scc = Vec::new();
                loop {
                    let w = stack.pop().unwrap();
                    in_stack[w] = false;
                    scc.push(w);
                    if w == u { break; }
                }
                scc_list.push(scc);
            }
        }

        // Build a DFS spanning tree over 'graph' restricted to 'nodes'
        fn build_spanning_tree_edges(
            &self,
            start: usize,
            graph: &[Vec<usize>],
            nodes: &HashSet<usize>,
        ) -> Vec<(usize, usize)> {
            let mut edges = Vec::new();
            let mut visited: HashSet<usize> = HashSet::new();
            let mut st = vec![start];
            visited.insert(start);

            while let Some(node) = st.pop() {
                for &nb in &graph[node] {
                    if nodes.contains(&nb) && !visited.contains(&nb) {
                        edges.push((node, nb));
                        visited.insert(nb);
                        st.push(nb);
                    }
                }
            }
            edges
        }

        // Tarjan's SCC (O(V+E))
        pub fn find_sccs(&self) -> Vec<Vec<usize>> {
            let mut disc = vec![-1i32; self.v];
            let mut low = vec![-1i32; self.v];
            let mut in_stack = vec![false; self.v];
            let mut stack = Vec::new();
            let mut scc_list = Vec::new();
            let mut timer = 0i32;

            for i in 0..self.v {
                if disc[i] == -1 {
                    self.tarjan_dfs(i, &mut disc, &mut low, &mut stack, &mut in_stack, &mut timer, &mut scc_list);
                }
            }
            scc_list
        }

        // Minimal SCC Edge Reduction (O(V+E))
        pub fn minimize_edges_in_scc(&self, scc: &[usize]) -> Vec<(usize, usize)> {
            if scc.is_empty() { return Vec::new(); }
            let nodes: HashSet<usize> = scc.iter().copied().collect();
            let mut essential_edges = Vec::new();
            let forward_tree = self.build_spanning_tree_edges(scc[0], &self.adj, &nodes);
            let reverse_tree = self.build_spanning_tree_edges(scc[0], &self.rev_adj, &nodes);
            essential_edges.extend(forward_tree);
            essential_edges.extend(reverse_tree);
            essential_edges
        }

        fn edge_count(&self) -> usize { self.adj.iter().map(|v| v.len()).sum() }

        pub fn reduce_edges_parallel(&self) -> Vec<(usize, usize)> {
            // Always compute SCCs sequentially for correctness and determinism
            let sccs = self.find_sccs();
            if self.verbose { println!("Found {} SCC(s).", sccs.len()); }

            // Small-N fallback to avoid parallel overhead
            let small = self.v <= 1024 || sccs.len() <= 2;
            if small {
                let mut reduced_edges = Vec::new();
                for scc in &sccs { reduced_edges.extend(self.minimize_edges_in_scc(scc)); }
                if self.verbose { println!("Reduced SCC edges: {}", reduced_edges.len()); }
                return reduced_edges;
            }

            // Parallel over SCCs but collect in original order for determinism
            let per_scc: Vec<Vec<(usize, usize)>> = sccs
                .par_iter()
                .map(|scc| self.minimize_edges_in_scc(scc))
                .collect(); // IndexedParallelIterator preserves order

            let total_len: usize = per_scc.iter().map(|v| v.len()).sum();
            let mut reduced_edges = Vec::with_capacity(total_len);
            for v in per_scc { reduced_edges.extend(v); }
            if self.verbose { println!("Reduced SCC edges: {}", reduced_edges.len()); }
            reduced_edges
        }
    }
}
